library/alloc/src/collections/fixed_queue/mod.rs

```diff
@@ -0,0 +1,288 @@
+            // shift head ptr on collision
+            self.head = (self.head + 1) % N;
+        } else {
+            // increase len if no collision
+            self.len += 1;
+        }
+        return overwritten;
+    }
+
+    /// Removes and returns the oldest element from the queue.
+    #[inline]
+    #[unstable(feature = "fixed_queue", issue = "126204")]
+    pub fn pop(&mut self) -> Option<T>
+    where
+        Option<T>: Copy,
+    {
+        if self.len == 0 {
+            return None;
+        }
+        let popped = self.buffer[self.head].take();
+        self.head = (self.head + 1) % N;
+        self.len -= 1;
+        popped
+    }
+
+    /// Converts the queue into its array equivalent.
+    #[unstable(feature = "fixed_queue", issue = "126204")]
+    pub fn to_option_array(self) -> [Option<T>; N]
+    where
+        Option<T>: Copy,
+    {
+        let mut arr: [Option<T>; N] = [None; N];
+        for i in 0..N {
+            arr[i] = self.buffer[(self.head + i) % N];
+        }
+        arr
+    }
+
+    /// Converts the queue into its vec equivalent.
+    #[unstable(feature = "fixed_queue", issue = "126204")]
+    pub fn to_vec(self) -> Vec<T>
+    where
+        T: Copy,
+    {
+        let mut vec: Vec<T> = Vec::new();
+        for i in 0..N {
+            if let Some(e) = self.buffer[(self.head + i) % N] {
+                vec.push(e);
+            }
+        }
+        vec
+    }
+}
+
+#[unstable(feature = "fixed_queue", issue = "126204")]
+impl<T, const N: usize> From<[T; N]> for FixedQueue<T, N> {
+    /// Creates a FixedQueue from a fixed size array.
+    fn from(array: [T; N]) -> Self {
+        FixedQueue::with(array.map(Some), 0, 0, N)
+    }
+}
+
+#[unstable(feature = "fixed_queue", issue = "126204")]
+impl<T: Copy, const N: usize> From<&[T; N]> for FixedQueue<T, N> {
+    /// Creates a FixedQueue from a fixed size slice.
+    fn from(array: &[T; N]) -> Self {
+        FixedQueue::with(array.map(Some), 0, 0, N)
+    }
+}
+
+#[unstable(feature = "fixed_queue", issue = "126204")]
+impl<T: Copy, const N: usize> From<&[T]> for FixedQueue<T, N> {
+    /// Creates a FixedQueue from an unsized slice. Copies a maximum of N
+    /// elements of the slice, and a minimum of the slice length into the
+    /// queue. {[0, 0], 0} - [0, 0]
+    fn from(array: &[T]) -> Self {
+        let mut buf: [Option<T>; N] = [None; N];
+        let length = N.min(array.len());
+        for i in 0..length {
+            buf[i] = Some(array[i]);
+        }
+        FixedQueue::with(buf, 0, array.len() - 1, length)
+    }
+}
+
+#[unstable(feature = "fixed_queue", issue = "126204")]
+impl<T: PartialEq, const N: usize> PartialEq for FixedQueue<T, N> {
+    /// This method tests if a FixedQueue is equal to another FixedQueue.
+    fn eq(&self, other: &FixedQueue<T, N>) -> bool {
+        if other.len != self.len {
+            return false;
+        }
+        (0..N).all(|x| self.buffer[(self.head + x) % N] == other.buffer[(other.head + x) % N])
```

Comment:
Like my remarks on `clear`: this should take advantage of its knowledge of where the head and tail are instead of destroying the entire cache if N is large.